File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dy = [-1,1,0,0]
dx = [0,0,-1,1]
def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    q = deque()
    q.append((0, 0))
    
    visited = [[-1]*m for _ in range(n)]
    visited[0][0] = 1
    
    while q:
        ey, ex = q.popleft()
        
        for i in range(4):
            ny, nx = ey + dy[i], ex + dx[i]
            
            if 0<=ny<n and 0<=nx<m:
                if maps[ny][nx] == 1 and visited[ny][nx] == -1:
                    visited[ny][nx] = visited[ey][ex] + 1
                    q.append((ny, nx))
    return visited[n-1][m-1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        ArrayDeque<int[]> q = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][m];
        q.offer(new int[]{0,0,1});
        visited[0][0] = true;
        
        int[] dy = {-1,1,0,0};
        int[] dx = {0,0,-1,1};
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            
            if (cur[0]==n-1 && cur[1]==m-1) return cur[2];
            
            for (int i=0;i<4;i++) {
                int ny = cur[0] + dy[i];
                int nx = cur[1] + dx[i];
                
                if (0<=ny && ny<n && 0<=nx && nx<m) {
                    if (!visited[ny][nx] && maps[ny][nx]==1) {
                        visited[ny][nx] = true;
                        q.offer(new int[]{ny, nx, cur[2]+1});
                    }
                }
            }
        }
                       
        return -1;
    }
}